spec.md

```diff
@@ -1586,12 +1760,192 @@ en <= Z_valid
+  input clock:  Clock
+  output a : Probe<{x: UInt, y: UInt}> ; read-only ref. to wire 'p'
+  output b : RWProbe<UInt> ; force-able ref. to node 'q', inferred width.
+  output c : Probe<UInt<1>> ; read-only ref. to register 'r'
+  output d : RWProbe<UInt<2>>[4] ; vector of ref.'s to memory data in 'm'
+  output e : Probe<Clock> ; ref. to input clock port
+
+  wire p : {x: UInt, flip y : UInt}
+  export probe(p) as a ; probe is passive
+  node q = UInt<1>(0)
+  export rwprobe(q) as b
+  reg r: UInt, clock
+  export probe(r) as c
+  mem m:
+    data-type => UInt<5>
+    depth => 4
+    ; ...
+    read-under-write => undefined
+
+  export rwprobe(m) as d
+  export probe(clock) as e
+```
+
+Exporting to a field within a bundle or other statically known sub-element of
+an aggregate is allowed, for example:
+```firrtl
+module Foo:
+  input x : UInt
+  output y : {x: UInt, p: Probe<UInt>}
+  output z : Probe<UInt>[2]
+
+  wire p : UInt
+  p <= x
+  y.x <= p
+
+  export probe(p) as y.p
+  export probe(p) as z[0]
+  export probe(p) as z[1]
+```
+
+`RWProbe`{.firrtl} references to ports are not allowed on public-facing modules.
+
+#### Probes and Passive Types
+
+While `Probe`{.firrtl} inner types are passive, the type of the exported expression
+is not required to be:
+
+```firrtl
+module Foo :
+  input x : {a: UInt, flip b: UInt}
+  output y : {a: UInt, flip b: UInt}
+  output xp : Probe<{a: UInt, b: UInt}> ; passive
+
+  wire p : {a: UInt, flip b: UInt} ; p is not passive
+  export probe(p) as xp
+  p <= x
+  y <= p
+```
+
+#### Exporting References to Nested Declarations
+
+Nested declarations (see [@sec:nested-declarations]) may be exported:
+
+```firrtl
+module RefProducer :
+  input a : UInt
+  input en : UInt<1>
+  output thereg : Probe<UInt>
+
+  when en :
+    reg myreg : UInt, clk
+    myreg <= a
+    export probe(myreg) as thereg
+```
+
+### Forward
+
+The forward statement is similar to export but forwards an existing reference
+to the specified target.
+
+This can be used to pass a child module's reference further up the hierarchy:
+
+```firrtl
+module Foo :
+  output p : Probe<UInt>
+  ; ...
+
+module Forward :
+  output p : Probe<UInt>
+
+  inst f of Foo
+  forward f.p as p
+```
+
+The forwarded expression and the export target must both be static expressions,
+and like the export statement do not participate in last-connect semantics.
+
+Forward statements may narrow a probe of an aggregate to a sub-element using
+static expression:
+
+```firrtl
+module Foo :
+  output p : Probe<UInt[2]>[2]
+  ; ...
+
+module Forward :
+  output p : Probe<UInt>
+
+  inst f of Foo
+  forward f.p[0][1] as p
+```
+
+### Force and Release
+
+To override existing drivers for a `RWProbe`{.firrtl}, the `force`{.firrtl}
+statement is used.  Force statements are simulation-only constructs and may not
+be supported by all implementations.
+
+* `force(clock, condition, refDst, value)`{.firrtl}: always posedge clock
+* `force(refDst, value)`{.firrtl}: initial
```

Comment:
Sounds great to me! Thanks!

Reply:
What happens if there are multiple `force`/`release` statements that are active at the same clock event?

Reply:
Seems like sticking to last connect semantics makes for a uniform model. However, if we are doing last-connect, then that should be explicitly stated in the spec.

Reply:
> Could we error instead (generate the exclusivity checking ourselves) ?

You could add a runtime assertion. A sound and complete compile time check might be impossible.

Reply:
Force overwrites the existing drivers (notice that since FIRRTL already requires all signals to be driven somehow, any force target already has a driver.  In other words: force is not a write and cannot be the normal driver), and has other behavior (including interactions with `release` and other `force` statements, as you say they're procedural and replace the previous `force`/assigns/drivers).  An important goal here is to also allow force-able targets to be used from raw verilog as well, which undermines the notion of "how about we have our own idea of force-like" such as having a "is/which force is active" signal, as well as makes it difficult to ensure the exclusivity (since these targets may be consumed by uses that do not use our protocol).

As for checking exclusivity ourselves, that sounds challenging to make work in general but especially across separately compiled units-- not all forces are `initial force x = y`-- that is, they may be enabled/disabled on-the-fly in a way that I really don't think we want to try to synthesize tracking / checking for.  It's my understanding that generally 1)force/release are not considered appropriate for design code, only for debugging/verification/testing 2)are not usually synthesizable, but don't have direct data/references for (2).

A FIRRTL-specific operation might be worth thinking about (where it's compatible with these, or using a FIRRTL-only reference/probe type), but I'm unsure of the use case there.  Happy to discuss/pursue separately if this seems useful to folks.

In response to the earlier question, when in doubt assume `force`/`release` behave as defined in Verilog.  See SV 2012 10.6 for more information, although I personally would encourage anyone interested to also explore how they behave in various simulators as it can be surprising and not consistently implemented.

I'll be updating the force/release section soon (hopefully today?) to clarify this and the other feedback (thanks!).

Reply:
> An important goal here is to also allow force-able targets to be used from raw verilog as well

In that case you would not use the `release` or `force` statement though, right? Since the force would come from outside FIRRTL. This case should probably be specified separately under a "FFI" or "Verilog-interop" section.

> As for checking exclusivity ourselves, that sounds challenging to make work in general but especially across separately compiled units

Are you saying that there could be `force`s from different modules to the same signal? If so, how are they resolved in a deterministic manner?

> A FIRRTL-specific operation might be worth thinking about

Isn't this PR about exactly that? I.e., FIRRTL specific statements and reference types that allow you to use `force` and `release` from FIRRTL?

> See SV 2012 10.6 for more information

Would you mind adding a description of the behavior to the FIRRTL spec directly? Referencing the SystemVerilog standard might raise more questions than it answers since FIRRTL has a different execution model.

> explore how they behave in various simulators as it can be surprising and not consistently implemented.

That indicates to me that FIRRTL should properly define the correct behavior such that the compiler can generate fixes and workarounds for broken simulators. A FIRRTL circuit should behave exactly the same no matter which simulator is used. Otherwise, our eco-system won't be any better than SystemVerilog where everyone seems to specify that their tests only work with ModelSim or VCS, or some other specific simulator version. We can and should be better than this (and have been so far).

I would like to encourage everyone to keep FIRRTL somewhat backend agnostic. I.e., we should be able to eventually have a CIRCT based `treadle` replacement which correctly implements all FIRRTL features. If we get in the habit of just pointing to the SystemVerilog spec, then FIRRTL becomes just a SystemVerilog frontend instead of a general hardware description IR.